### PR TITLE
Rearrange output folders

### PR DESCRIPTION
**Pull request by dsplaisted**
- Rearrange artifact (output) folder
- Better separation between building and testing (separate stages)
- This makes it possible to run a full build twice in a row without failing

### CODE REVIEW

File: build_projects/dotnet-cli-build/DotNetMSBuildTool.cs [hunk 1]
```diff
@@ -40,5 +42,10 @@ private string GetVerbosityArg()
 
             return null;
         }
+
+        private string GetAdditionalParameters()
+        {
+            return AdditionalParameters;
+        }
```
Comment (nguerrera) [inline build_projects/dotnet-cli-build/DotNetMSBuildTool.cs]: What's the point of this method? Why not just use the property?

File: build/OutputDirectories.props [hunk 2]
```diff
@@ -1,20 +1,29 @@
...
+
+      <!-- BaseOutputDirectory is the root output path for this stage and RID
+           BaseOutputPath is the project-specific output folder that each project builds into.
+           This is not confusing at all. -->
+      <BaseOutputDirectory>$(RepoRoot)/artifacts/stage$(CliBuildStage)/$(Rid)</BaseOutputDirectory>
+      <BaseOutputPath>$([System.IO.Path]::GetFullPath('$(BaseOutputDirectory)/bin/$(MSBuildProjectName)'))</BaseOutputPath>
+
+      <BaseIntermediateOutputPath>$([System.IO.Path]::GetFullPath('$(BaseOutputDirectory)/obj/$(MSBuildProjectName)'))</BaseIntermediateOutputPath>
+      <OutputDirectory>$(BaseOutputDirectory)/dotnet</OutputDirectory>
+      <Stage2WithBackwardsCompatibleRuntimesOutputDirectory>$(BaseOutputDirectory)/dotnetWithBackwardsCompatibleRuntimes</Stage2WithBackwardsCompatibleRuntimesOutputDirectory>
```
Comment (nguerrera) [inline build/OutputDirectories.props]: I see these are also set in signing.proj. Will the mismatched values break signing? Can you get rid of the duplication?

File: build/OutputDirectories.props [hunk 3]
```diff
@@ -1,20 +1,29 @@
...
+      <SymbolsDirectory>$(BaseOutputDirectory)/symbols</SymbolsDirectory>
       <RoslynDirectory>$(SdkOutputDirectory)/Roslyn</RoslynDirectory>
       <FSharpDirectory>$(SdkOutputDirectory)/FSharp</FSharpDirectory>
-      <CompilationDirectory>$(BaseOutputDirectory)/stage2compilation</CompilationDirectory>
+      <CompilationDirectory>$(BaseOutputDirectory)/compilation</CompilationDirectory>
       <IntermediateDirectory>$(BaseOutputDirectory)/intermediate</IntermediateDirectory>
       <PackagesDirectory>$(BaseOutputDirectory)/packages</PackagesDirectory>
       <SharedFrameworkPublishDirectory>$(IntermediateDirectory)/sharedFrameworkPublish</SharedFrameworkPublishDirectory>
       <!-- The 'AspNetRuntimePackageStorePublishDirectory' needs to be very short due to path length constraints on Win8.1 -->
+      <!-- Will this still work now that the stage has been added to the path? -->
```
Comment (nguerrera) [inline build/OutputDirectories.props]: Hmm

File: build_projects/dotnet-cli-build/DotNetMSBuildTool.cs (hunk 1 above)
Comment (dsplaisted) [reply build_projects/dotnet-cli-build/DotNetMSBuildTool.cs]: I was trying to follow the pattern I saw in another similar class.  I've updated it now.

**Comment by wli3** [comment]
@dsplaisted windows tests are failing https://ci.dot.net/job/dotnet_cli/job/master/job/debug_windows_nt_x86_prtest/1424/testReport/

File: build/OutputDirectories.props [hunk 4]
```diff
@@ -1,23 +1,35 @@
...
-      <BaseOutputDirectory>$(RepoRoot)/artifacts/$(Rid)</BaseOutputDirectory>
-      <OutputDirectory>$(BaseOutputDirectory)/stage2</OutputDirectory>
-      <Stage2WithBackwardsCompatibleRuntimesOutputDirectory>$(BaseOutputDirectory)/stage2WithBackwardsCompatibleRuntimes</Stage2WithBackwardsCompatibleRuntimesOutputDirectory>
+      <CliBuildStage Condition="'$(CliBuildStage)' == '' And '$(CliOuterBuildStage)' != ''">$([MSBuild]::Add($(CliOuterBuildStage), '1'))</CliBuildStage>
+      <CliBuildStage Condition="'$(CliBuildStage)' == ''">2</CliBuildStage>
+
+      <!-- BaseOutputDirectory is the root output path for this stage and RID
+           BaseOutputPath is the project-specific output folder that each project builds into.
+           This is not confusing at all. -->
+      <BaseOutputDirectory>$(RepoRoot)/out/$(CliBuildStage)/$(Rid)</BaseOutputDirectory>
```
Comment (eerhardt) [inline build/OutputDirectories.props]: Can the `out` directory be renamed to `bin` to align with the rest of the dotnet repos?

Comment (dsplaisted) [reply build/OutputDirectories.props]: @eerhardt There are also `bin` directories further down under `out` (ie `out\2\win10-x64\bin\<ProjectName>`), and the intermediate (`obj`) folders are traditionally a peer to `bin`, but they are also under the `out` folder.  Given that, would you still prefer `bin` for the top-level output directory?

Comment (eerhardt) [reply build/OutputDirectories.props]: Yes, it is consistent with the rest of the repos.  `corefx`, `coreclr`, `core-setup`, `standard`, `sdk`, `BuildTools` all use `bin` at the root of the repo.

These repos all make the intermediate output folder under `bin/obj`.

The term `artifacts` mostly came from ASP.NET and aligns with their repos.

If we are going to change this folder's name, we should change it to align with the majority of the `dotnet` repos.  In general, we need to be more consistent across our repos to make it easier for other team members to contribute across repos.

File: build/OutputDirectories.props (hunk 2 above)
Comment (nguerrera) [reply build/OutputDirectories.props]: https://github.com/dotnet/cli/blob/8ef162fd13005a56edf8729cdd91a76eaac0309d/build/Signing.proj#L16-L20

File: build/OutputDirectories.props (hunk 3 above)
Comment (nguerrera) [reply build/OutputDirectories.props]: Delete. Should be just as short now, right?

**Comment by dsplaisted** [comment]
@dotnet-bot test Debian8.2 x64 Debug Build

**Comment by dsplaisted** [comment]
@dotnet-bot test CentOS7.1 x64 Debug Build

File: test/dotnet.Tests/dotnet.Tests.csproj [hunk 5]
```diff
@@ -22,11 +22,35 @@
...
+      </ExistingVersionFileContent>
+
+      <ShouldOverwriteVersionFile>false</ShouldOverwriteVersionFile>
+      <ShouldOverwriteVersionFile
+        Condition=" '$(ExistingVersionFileContent.Trim())' != '$(VersionFileContent.Trim())' ">true</ShouldOverwriteVersionFile>
+    </PropertyGroup>
+    
+    <WriteLinesToFile File="$(ExpectedVersionFileInIntermediateFolder)"
+                      Lines="$(VersionFileContent)"
+                      Condition=" '$(ShouldOverwriteVersionFile)' == 'true' "
```
Comment (nguerrera) [inline test/dotnet.Tests/dotnet.Tests.csproj]: There's a Write Only When Different option on Write Lines To File now that you can use to simplify this

**Pull request merged**